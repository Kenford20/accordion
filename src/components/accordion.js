import React from 'react';
import DealHeader from './deal-header';

export default class Accordion extends React.Component {
    constructor(props) {
      super(props);
      
      this.state = {
        deals: {
          "Local": [10, 20, 30],
          "Getaway": [40, 50, 60],
          "Goods": [70, 80, 90],
          "Groupon": [150]
        },
        grandTotal: 0
      }
    }
    
    componentDidMount() {
      let initialGrandTotal = 0;
      Object.keys(this.state.deals).map(deal => {
        this.state.deals[deal].map(price => initialGrandTotal += parseInt(price));
      });
      
      this.setState({ grandTotal: initialGrandTotal });
    }
    
    addDeal = (dealType) => {
      const randomNum = Math.floor(Math.random()*100);
      this.setState(prevState => {
        return { 
          grandTotal: prevState.grandTotal + randomNum,
          deals: {
            ...prevState.deals,
            [dealType]: prevState.deals[dealType].concat(randomNum)
          }
        }
      });
    }
    
    render() {
      const { deals, grandTotal } = this.state;
      return (
        <div id="accordion">
          {
            Object.keys(deals).map((type, index) => (
              <DealHeader 
                header={type}
                prices={deals[type]}
                addDeal={this.addDeal}
                key={index}
              />
            ))
          }
          <div id="grand-total">
            <div>Grand Total: <span>${grandTotal}</span></div>
          </div>
        </div>
      );
    };
  }