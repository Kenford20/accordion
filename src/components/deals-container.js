import React from 'react';

export const DealsContainer = ({ deal, header, addDeal }) => {
    return (
      <div className="deals-container" data-testid="deals-container">      
        {
          deal.map((price, index) => (
            <div className="deal" key={index} data-testid="deal">
              <span>Deal {++index}</span>
              <span>${price}</span>
            </div>
          ))
        }
        <div className="add-deal">
          <span data-testid="add-deal-btn" onClick={() => addDeal(header)}>Add Deal</span>
        </div> 
      </div>
    );
}