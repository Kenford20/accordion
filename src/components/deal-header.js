import React, { useState, useEffect } from 'react';
import { DealsContainer } from './deals-container';

export const DealHeader = ({ prices, header, addDeal }) => {
  const [dealsTotal, setDealsTotal] = useState(0);
  //const [showDeals, toggleShowDeals] = useState(true);

  useEffect(() => {
    setDealsTotal(prices.reduce((a,b) => a+b));
  }, [prices]);

  const toggleDeals = (e) => {
    // toggleShowDeals(!showDeals);
    
    // logic below is used instead of setstate above to allow for sliding animation
    // handle user clicks on the text/total (aka the spans) to still properly toggle the deals container
    const targetNode = e.target.parentNode.classList.contains('deal-header-containers') 
      ? e.target.parentNode.nextSibling 
      : e.target.nextSibling;
    
    if(targetNode.style.maxHeight !== '0px') {
      targetNode.style.maxHeight = '0px';
      targetNode.style.opacity = '0';
    } else {
      targetNode.style.maxHeight = '2000px';
      targetNode.style.opacity = '100';
    }
  }
  
  return (
    <React.Fragment>
      <div className="deal-header-containers" data-testid="deal-header" onClick={(e) => toggleDeals(e)}>
        <span data-testid="deals-type">{header} Deals</span>
        <span data-testid="deals-total">Total: ${dealsTotal}</span>
      </div>

      <DealsContainer 
        deal={prices} 
        header={header} 
        addDeal={addDeal}
      /> 
    </React.Fragment>
  );
};
