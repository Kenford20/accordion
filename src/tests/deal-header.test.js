import React from "react";
import { DealHeader } from "../components/deal-header";
import { render, cleanup, fireEvent } from "@testing-library/react";
import "jest-dom/extend-expect";

afterAll(cleanup);

const fakePrices = [10, 20, 30, 40, 50];
const header="Groupon";
const { container, getByTestId } = render(
    <DealHeader header={header} prices={fakePrices}/>
);

it("renders the deal header", () => {
    expect(container).toMatchSnapshot();
});

it("should render the correct deal type in the header", () => {
    const dealType = getByTestId("deals-type");
    expect(dealType).toHaveTextContent(header);
});

it("should display the correct total price of deals", () => {
    const totalPriceNode = getByTestId("deals-total");
    const totalPrice = fakePrices.reduce((a,b) => a+b);

    expect(totalPriceNode).toHaveTextContent(totalPrice);
});

it("should toggle the deals container's visibility when it is clicked", () => {
    const dealsContainer = getByTestId("deals-container");
    const dealHeader = getByTestId("deal-header");

    expect(dealsContainer).toBeVisible();
    fireEvent.click(dealHeader);
    expect(dealsContainer).not.toBeVisible();
    fireEvent.click(dealHeader);
    expect(dealsContainer).toBeVisible();
});




