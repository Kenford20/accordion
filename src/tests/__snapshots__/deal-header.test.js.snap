// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the deal header 1`] = `
<div>
  <div
    class="deal-header-containers"
    data-testid="deal-header"
  >
    <span
      data-testid="deals-type"
    >
      Groupon
       Deals
    </span>
    <span
      data-testid="deals-total"
    >
      Total: $
      150
    </span>
  </div>
  <div
    class="deals-container"
    data-testid="deals-container"
  >
    <div
      class="deal"
      data-testid="deal"
    >
      <span>
        Deal 
        1
      </span>
      <span>
        $
        10
      </span>
    </div>
    <div
      class="deal"
      data-testid="deal"
    >
      <span>
        Deal 
        2
      </span>
      <span>
        $
        20
      </span>
    </div>
    <div
      class="deal"
      data-testid="deal"
    >
      <span>
        Deal 
        3
      </span>
      <span>
        $
        30
      </span>
    </div>
    <div
      class="deal"
      data-testid="deal"
    >
      <span>
        Deal 
        4
      </span>
      <span>
        $
        40
      </span>
    </div>
    <div
      class="deal"
      data-testid="deal"
    >
      <span>
        Deal 
        5
      </span>
      <span>
        $
        50
      </span>
    </div>
    <div
      class="add-deal"
    >
      <span
        data-testid="add-deal-btn"
      >
        Add Deal
      </span>
    </div>
  </div>
</div>
`;
